accepted/0000-sbom-command.md

```diff
@@ -0,0 +1,461 @@
+  "versionInfo": "4.3.4",
+  "downloadLocation": "https://registry.npmjs.org/debug/-/debug-4.3.4.tgz",
+  "filesAnalyzed": false,
+  "externalRefs": [
+    {
+      "referenceCategory": "PACKAGE-MANAGER",
+      "referernceType": "npm",
+      "referenceLocator": "debug@4.3.4"
+    },
+    {
+      "referenceCategory": "PACKAGE-MANAGER",
+      "referernceType": "purl",
+      "referenceLocator": "pkg:npm/debug@4.3.4"
+    }
+  ],
+  "checksums": [
+    {
+      "algorithm": "SHA512",
+      "checksumValue": "3d15851ee494dde0ed4093ef9cd63b..."
+    }
+  ]
+}
+```
+
+The example record shown above conforms to version 2.3 of the SPDX [Package](https://spdx.github.io/spdx-spec/v2.3/package-information/) specification.
+
+The <code>[downloadLocation](https://spdx.github.io/spdx-spec/v2.3/package-information/#77-package-download-location-field)</code> field will report the resolved location calculated by Arborist. This may point to a package registry, or a git URL for dependencies referenced directly from git. The top-level project will specify a value of <code>NOASSERTION</code> for the download location as this information is not available.
+
+All packages will specify a `false` value for the <code>[filesAnaylzed](https://spdx.github.io/spdx-spec/v2.3/package-information/#78-files-analyzed-field)</code> field to signal that no assertions are being made about the files contained within a package.
+
+The <code>[externalRefs](https://spdx.github.io/spdx-spec/v2.3/package-information/#721-external-reference-field)</code> field will contain two <code>[PACKAGE-MANAGER](https://spdx.github.io/spdx-spec/v2.3/external-repository-identifiers/#f3-package-manager)</code> references, one using the <code>[npm](https://spdx.github.io/spdx-spec/v2.3/external-repository-identifiers/#f32-npm)</code> reference type and the other using the <code>[purl](https://spdx.github.io/spdx-spec/v2.3/external-repository-identifiers/#f35-purl)</code> reference type.
+
+
+## Unresolved Questions and Bikeshedding
+
+* There’s a case to be made that SBOM generation should just be a flag on the existing `npm-ls` command – an SBOM is really just a different format of the output already generated by `npm-ls` (in fact, the CycloneDX tool described above parses the output from `npm-ls` in order to generate its SBOM). However, many of the arguments supported by `npm-ls` don’t really make sense in the context of SBOM generation (package spec filtering, depth, etc…) so it feels like a bit of an awkward fit. \
+ \
+Making it a distinct command allows us to add SBOM-specific features in the future like a `--sign` flag which could be used to generate a signed SBOM. \
+ \
+_Recommendation: Add a distinct command for generating an SBOM._
+ 
+* Does `npm-sbom` command have a notion of a “default” SBOM format? Do we give preference to one of CycloneDX/SPDX or do we remain totally neutral (possibly at the expense of DX)? \
+ \
+_Recommendation: Default to SPDX if no format is specified._
```

Comment:
disagree. better be neutral.

Reply:
Also disagree. If you look at the download stats from the various CycloneDX Javascript implementations, it's clear that CycloneDX is what most people are using today. But, I'd rather see npm be neutral and explicitly require the user to specify a format.

Reply:
Was trying to err on the side of developer experience (i.e. fewer required flags) but also understand the case for format neutrality. 

Would appreciate someone from the npm team to weigh-in on the most idiomatic way to represent this -- I don't believe npm has any notion of "required" flags, so it would just need to be an argument to command:

```
npm sbom spdx
```

That doesn't feel quite right 🤔

Also would be nice if developers could set this value once (either via env var, .npmrc config, or package.json entry) to alleviate the burden of having to always specify this value.

Reply:
Many orgs will be producing both formats. Having something set in .npmrc also feels wrong as the same build system may likely build different projects requiring different formats. One option is to automatically produce both formats with `npm sbom`.

Reply:
Right now everyone working on SBOMs knows and cares about the formats, that's not going to be the case for long. Making people pick and look this up is going to hurt long-term adoption of SBOMs. I'm in favor of just picking a default to avoid requiring choices.

If you care which SBOM you use, it's the same amount of work. If you don't, it's less mental energy, and the people that care need the least friction.

Reply:
That seems the better choice than trying to pick a “winner”.

Reply:
CycloneDX has some [Recognized file patterns](https://cyclonedx.org/specification/overview/#recognized-file-patterns), like `*.cdx.json`

I know it is common for SPDX files to match `*.spdx.json`, but i did not find the docs for it.

So npm's SBOM command could result in creating two files at once: `sbom.cdx.json` and `sbom.spdx.json`

Reply:
I think it's important to be able to write the generated SBOM to stdout (npm users expect to be able to pipe/redirect output from the various commands). This would seem to eliminate the "output both formats" option (either writing directly to a file or emitting both SBOMs to stdout).

After talking w/ the npm dev team, it seems like creating a mandatory `--sbom-format` flag is do'able. This has the added benefit of automatically enabling the ability to set this value via env var or npmrc config for use cases where developers know they always want a specific format.

I'm going to update this bullet item with this new recommendation.

Reply:
Have updated the recommendation to be neutral and use a mandatory option to force the user to make an explicit format selection.